src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ForeachToFor
+{
+    internal abstract class AbstractForeachToForCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        private const string get_Count = nameof(get_Count);
+        private const string get_Item = nameof(get_Item);
+
+        private const string Length = nameof(Length);
+        private const string Count = nameof(Count);
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            new string[] { typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName }.ToImmutableArray();
+
+        protected abstract SyntaxNode GetForeachStatement(SyntaxToken token);
+        protected abstract (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode foreachStatement);
+        protected abstract void ConvertToForStatement(SemanticModel model, ForeachInfo info, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var foreachStatement = GetForeachStatement(token);
+            if (foreachStatement == null)
+            {
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var foreachInfo = GetForeachInfo(semanticFact, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForeachToForCodeAction(
+                    FeaturesResources.Convert_foreach_to_for,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected string CreateUniqueMethodName(SemanticModel model, SyntaxNode contextNode, string baseName)
+        {
+            Contract.ThrowIfNull(contextNode);
+            Contract.ThrowIfNull(baseName);
+
+            return NameGenerator.GenerateUniqueName(baseName, string.Empty,
+               name => model.LookupSymbols(contextNode.SpanStart, /*container*/null, name).Length == 0);
+        }
+
+        protected T AddElasticAnnotation<T>(T node, SyntaxTrivia elasticTrivia) where T : SyntaxNode
+        {
+            var first = node.GetFirstToken();
+            node = node.ReplaceToken(first, first.WithPrependedLeadingTrivia(elasticTrivia));
+
+            var last = node.GetLastToken();
+            return node.ReplaceToken(last, last.WithAppendedTrailingTrivia(elasticTrivia));
+        }
```

Comment:
you can already do   node.WithPrependedLeadingTrivia(elastictrivia).WithAppendedTrailingTrivia(...)
you don't need to do the explicit replace token calls.

--

Also, not sure why this method is needed.

Reply:
It's needed because you parsed out a string yourself.  Please don't do that.  Just construct the tree normally.

Reply:
ah. we had that extension method?